hail/python/hail/methods/relatedness/pc_relate_ts.py

```diff
@@ -0,0 +1,507 @@
+        Default given by :meth:`.BlockMatrix.default_block_size`.
+    include_self_kinship: :obj:`bool`
+        If ``True``, include entries for an individual's estimated kinship with
+        themselves. Defaults to ``False``.
+
+    Returns
+    -------
+    :class:`.Table`
+        A :class:`.Table` mapping pairs of samples to their pair-wise statistics.
+    """
+    assert (0.0 <= min_individual_maf <= 1.0), \
+        f"invalid argument: min_individual_maf={min_individual_maf}. " \
+        f"Must have min_individual_maf on interval [0.0, 1.0]."
+    mt = matrix_table_source('pc_relate_ts/call_expr', call_expr)
+
+    if k and scores_expr is None:
+        _, scores, _ = _hwe_normalized_blanczos(call_expr, k, compute_loadings=False, q_iterations=10)
+        scores_expr = scores[mt.col_key].scores
+    elif not k and scores_expr is not None:
+        analyze('pc_relate_ts/scores_expr', scores_expr, mt._col_indices)
+    elif k and scores_expr is not None:
+        raise ValueError("pc_relate_ts: exactly one of 'k' and 'scores_expr' "
+                         "must be set, found both")
+    else:
+        raise ValueError("pc_relate_ts: exactly one of 'k' and 'scores_expr' "
+                         "must be set, found neither")
+
+    scores_table = mt.select_cols(__scores=scores_expr) \
+        .key_cols_by().select_cols('__scores').cols()
+
+    n_missing = scores_table.aggregate(agg.count_where(hl.is_missing(scores_table.__scores)))
+    if n_missing > 0:
+        raise ValueError(f'Found {n_missing} columns with missing scores array.')
+
+    mt = mt.select_entries(__gt=call_expr.n_alt_alleles()).unfilter_entries()
+    mt = mt.annotate_rows(__mean_gt=agg.mean(mt.__gt))
+    mean_imputed_gt = hl.or_else(hl.float64(mt.__gt), mt.__mean_gt)
+
+    if not block_size:
+        block_size = BlockMatrix.default_block_size()
+
+    g = hl.experimental.mt_to_table_of_ndarray(mean_imputed_gt, block_size=block_size)
+    pc_scores = hl.nd.array(scores_table.collect(_localize=False).map(lambda x: x.__scores))
+
+    # g.shape is (n_variants, n_samples), pc_scores.shape is (n_samples, k)
+    n = g.take(1)[0].ndarray.shape[1]
+    sqrt_n = hl.nd.array([hl.sqrt(n)])
+
+    # Recover singular values, S0, as vector of column norms of pc_scores
+    S0 = (pc_scores ** hl.int32(2)).sum(0).map(lambda x: hl.sqrt(x))
+    # We want sqrt(n_samples) as first entry in S, for intercept in beta
+    S = hl.nd.hstack((sqrt_n, S0))._persist()
+    # Recover V from pc_scores with inv(S0)
+    V0 = (pc_scores * (1 / S0))._persist()
+    # First column in V needs all entries as 1/sqrt(n_samples), for intercept in beta
+    ones_normalized = hl.nd.full((V0.shape[0], 1), (1 / S[0]))
+    V = hl.nd.hstack((ones_normalized, V0))
+
+    nan = hl.float64(float("NaN"))
+
+    beta = g.select(ndarray=((1 / S) * V).T @ g.ndarray.T)
+    mu = beta.select(ndarray=((0.5 * V * S) @ beta.ndarray).T.map(lambda mu: hl.if_else(_bad_mu(mu, min_individual_maf), nan, mu)))
```

Comment:
It looks like from this point on, `beta` `S` and `V` aren't used again, only `mu`. Is that right?

You shouldn't need to actually materialize the `hstack`s. From some quick scratch work, it looks like you can equivalently use
```
g = g.annotate(af=g.ndarray.sum(1) / (2 * n))
mu = g.select(ndarray=g.af + 0.5 * g.ndarray @ V0 @ V0.T).map(lambda mu: hl.if_else(_bad_mu(mu, min_individual_maf), nan, mu)))
```
which feels more intuitive to me as well (mean center, compute PCs and project, then add the mean back). Happy to go through the math with you.